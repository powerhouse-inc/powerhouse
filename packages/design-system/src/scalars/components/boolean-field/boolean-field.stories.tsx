import type { Meta, StoryObj } from "@storybook/react";
import { withForm } from "../../lib/decorators.js";
import {
  getDefaultArgTypes,
  getValidationArgTypes,
  StorybookControlCategory,
} from "../../lib/storybook-arg-types.js";
import { BooleanField } from "./boolean-field.js";

const meta = {
  title: "Document Engineering/Scalars/Boolean Field",
  component: BooleanField,
  decorators: [withForm],
  argTypes: {
    ...getDefaultArgTypes({
      enabledArgTypes: {
        value: false,
      },
      valueControlType: "boolean",
      valueType: "boolean",
    }),

    value: {
      control: "inline-radio",
      options: ["indeterminate", true, false],
      description: "Current value of the input field",
      table: {
        type: { summary: "boolean" },
        category: StorybookControlCategory.DEFAULT,
      },
    },

    isToggle: {
      control: "boolean",
      description: "Whether the field is a toggle or checkbox",
      table: {
        type: { summary: "boolean" },
        defaultValue: { summary: "false" },
        category: StorybookControlCategory.COMPONENT_SPECIFIC,
      },
    },

    ...getValidationArgTypes({
      enabledArgTypes: {
        validators: false,
        showErrorOnBlur: false,
        showErrorOnChange: false,
      },
    }),
  },
  args: {
    name: "boolean-field",
  },
  tags: ["autodocs"],
  parameters: {
    layout: "centered",
    docs: {
      description: {
        component:
          "A boolean field component that can be used as a [checkbox](?path=/docs/document-engineering-data-entry-checkbox--readme) or [toggle](?path=/docs/document-engineering-data-entry-toggle--readme) depending on the `isToggle` prop.",
      },
    },
  },
} satisfies Meta<typeof BooleanField>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Checkbox: Story = {
  args: {
    label: "Checkbox Field",
    description: "This is a checkbox field",
    isToggle: false,
  },
};

export const Toggle: Story = {
  args: {
    label: "Toggle Field",
    description: "This is a toggle field",
    isToggle: true,
  },
};
