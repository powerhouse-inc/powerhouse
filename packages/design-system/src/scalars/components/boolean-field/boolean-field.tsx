import React from "react";
import type { CheckboxProps } from "../../../ui/components/data-entry/checkbox/checkbox.js";
import type { ToggleProps } from "../../../ui/components/data-entry/toggle/toggle.js";
import { CheckboxField } from "../fragments/checkbox-field/index.js";
import { ToggleField } from "../fragments/toggle-field/index.js";
export interface BooleanFieldProps
  extends Omit<CheckboxProps, "defaultValue" | "onChange" | "value">,
    Omit<ToggleProps, "onChange" | "value"> {
  isToggle?: boolean;
  onChange?: (value: boolean) => void;
  value?: boolean;
}

export const BooleanField = React.forwardRef<
  HTMLButtonElement,
  BooleanFieldProps
>(({ isToggle, onChange, ...props }, ref) => {
  const handleChange = (value: string | boolean) => {
    onChange?.(Boolean(value));
  };

  return isToggle ? (
    <ToggleField onChange={handleChange} ref={ref} {...props} />
  ) : (
    <CheckboxField onChange={handleChange} ref={ref} {...props} />
  );
});

BooleanField.displayName = "BooleanField";
