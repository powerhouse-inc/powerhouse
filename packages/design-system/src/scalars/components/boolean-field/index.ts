export { BooleanField } from "./boolean-field.js";
