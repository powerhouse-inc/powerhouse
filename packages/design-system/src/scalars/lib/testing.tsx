import { render, type RenderResult } from "@testing-library/react";
import { Form } from "../components/index.js";

export const renderWithForm = (
  children: React.ReactNode,
  onSubmit?: () => void,
): RenderResult => {
  return render(children, {
    wrapper: ({ children }) => (
      <Form onSubmit={onSubmit ?? (() => null)}>{children}</Form>
    ),
  });
};
