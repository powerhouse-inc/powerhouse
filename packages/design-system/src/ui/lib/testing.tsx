import { Form } from "@powerhousedao/design-system";
import type { RenderResult } from "@testing-library/react";
import { render } from "@testing-library/react";

export const renderWithForm = (
  children: React.ReactNode,
  onSubmit?: () => void,
): RenderResult => {
  return render(children, {
    wrapper: ({ children }) => (
      <Form onSubmit={onSubmit ?? (() => null)}>{children}</Form>
    ),
  });
};
