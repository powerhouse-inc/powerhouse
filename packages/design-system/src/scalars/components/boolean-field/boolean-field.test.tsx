import { renderWithForm } from "../../lib/testing.js";
import { BooleanField } from "./boolean-field.js";

describe("BooleanField", () => {
  const commonProps = {
    label: "Test Field",
    description: "Test description",
    value: true,
    required: true,
    disabled: false,
    errors: ["Test error"],
    warnings: ["Test warning"],
    className: "test-class",
    onChange: vi.fn(),
  };

  it("should render toggle field when isToggle is true", () => {
    const { container } = renderWithForm(
      <BooleanField name="test" {...commonProps} isToggle={true} />,
    );
    expect(container).toMatchSnapshot();
  });

  it("should render checkbox field when isToggle is false", () => {
    const { container } = renderWithForm(
      <BooleanField name="test" {...commonProps} isToggle={false} />,
    );
    expect(container).toMatchSnapshot();
  });

  it("should pass all props to ToggleField when isToggle is true", () => {
    const { getByRole, queryByText } = renderWithForm(
      <BooleanField name="test" {...commonProps} isToggle={true} />,
    );
    const toggle = getByRole("switch");
    expect(toggle).toBeChecked();
    expect(toggle).toBeEnabled();
    const toggleLabel = queryByText("Test Field");
    expect(toggleLabel).toHaveAttribute("for", toggle.id);
  });

  it("should pass all props to CheckboxField when isToggle is false", () => {
    const { getByRole, getByText } = renderWithForm(
      <BooleanField name="test" {...commonProps} isToggle={false} />,
    );
    const checkbox = getByRole("checkbox");
    expect(checkbox).toBeChecked();
    expect(checkbox).toBeEnabled();
    const checkboxLabel = getByText("Test Field");
    expect(checkboxLabel).toHaveAttribute("for", checkbox.id);
  });
});
