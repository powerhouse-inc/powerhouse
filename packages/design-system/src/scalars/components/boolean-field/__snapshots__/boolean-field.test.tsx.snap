// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BooleanField > should render checkbox field when isToggle is false 1`] = `
<div>
  <form
    id=":r3:"
    novalidate=""
  >
    <div
      class="flex flex-col gap-2"
    >
      <div
        class="group flex items-center space-x-2 test-class"
      >
        <button
          aria-checked="true"
          aria-invalid="false"
          aria-required="true"
          class="peer size-4 shrink-0 rounded border-input border shadow-sm shadow-black/[.04] ring-offset-background transition-shadow focus-visible:border-ring focus-visible:ring-ring focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:border-gray-700 disabled:data-[invalid=false]:data-[state=checked]:bg-gray-700 disabled:data-[invalid=false]:data-[state=indeterminate]:bg-gray-700 dark:disabled:data-[invalid=false]:data-[state=checked]:bg-gray-500 dark:disabled:data-[invalid=false]:data-[state=indeterminate]:bg-gray-500 data-[state]:border-gray-700 dark:data-[state]:border-gray-500 data-[state=checked]:bg-gray-900 data-[state=indeterminate]:bg-gray-900 dark:data-[state=checked]:bg-gray-400 dark:data-[state=indeterminate]:bg-gray-400 data-[state=checked]:text-slate-50 data-[state=indeterminate]:text-slate-50 dark:data-[state=checked]:text-gray-900 dark:data-[state=indeterminate]:text-gray-900 group-hover:border-gray-900 data-[state=checked]:group-hover:bg-gray-900 data-[state=indeterminate]:group-hover:bg-gray-900 dark:group-hover:border-slate-50 dark:data-[state=checked]:group-hover:bg-slate-50 dark:data-[state=indeterminate]:group-hover:bg-slate-50 data-[invalid=true]:data-[state]:!border-red-800 data-[invalid=true]:data-[state=checked]:!bg-red-800 data-[invalid=true]:data-[state=indeterminate]:!bg-red-800 dark:data-[invalid=true]:data-[state]:!border-red-800 dark:data-[invalid=true]:data-[state=checked]:!bg-red-800 dark:data-[invalid=true]:data-[state=indeterminate]:!bg-red-800 data-[invalid=true]:group-hover:!border-red-900 data-[invalid=true]:data-[state=checked]:group-hover:!bg-red-900 data-[invalid=true]:data-[state=indeterminate]:group-hover:!bg-red-900"
          data-invalid="false"
          data-state="checked"
          id=":r4:"
          role="checkbox"
          type="button"
          value="on"
        >
          <span
            class="flex items-center justify-center text-current"
            data-state="checked"
            style="pointer-events: none;"
          >
            <svg
              fill="currentcolor"
              height="9"
              viewBox="0 0 9 9"
              width="9"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M8.53547 0.62293C8.88226 0.849446 8.97976 1.3142 8.75325 1.66099L4.5083 8.1599C4.38833 8.34356 4.19397 8.4655 3.9764 8.49358C3.75883 8.52167 3.53987 8.45309 3.3772 8.30591L0.616113 5.80777C0.308959 5.52987 0.285246 5.05559 0.563148 4.74844C0.84105 4.44128 1.31533 4.41757 1.62249 4.69547L3.73256 6.60459L7.49741 0.840706C7.72393 0.493916 8.18868 0.396414 8.53547 0.62293Z"
                fill-rule="evenodd"
              />
            </svg>
          </span>
        </button>
        <input
          aria-hidden="true"
          checked=""
          name="test"
          required=""
          style="transform: translateX(-100%); position: absolute; pointer-events: none; opacity: 0; margin: 0px; width: 0px; height: 0px;"
          tabindex="-1"
          type="checkbox"
          value="on"
        />
        <label
          class="inline-flex items-center text-sm font-medium leading-[22px] text-gray-900 dark:text-gray-400 group-hover:text-gray-900 dark:group-hover:text-slate-50 group-hover:cursor-pointer"
          for=":r4:"
        >
          Test Field
          <span
            class="ml-1 text-gray-800 group-hover:text-gray-900 dark:text-gray-400 dark:group-hover:text-slate-50"
          >
            *
          </span>
          <button
            data-state="closed"
            type="button"
          >
            <div
              data-testid="icon-fallback"
              style="width: 16px; height: 16px;"
            />
          </button>
        </label>
      </div>
      <p
        class="mb-0 inline-flex items-center text-xs font-medium text-yellow-900 dark:text-yellow-900"
        data-type="warning"
      >
        Test warning
      </p>
    </div>
  </form>
</div>
`;

exports[`BooleanField > should render toggle field when isToggle is true 1`] = `
<div>
  <form
    id=":r0:"
    novalidate=""
  >
    <div
      class="flex flex-col gap-1 test-class"
      data-testid="custom-class"
    >
      <div
        class="flex items-center"
      >
        <button
          aria-checked="true"
          aria-labelledby=":r1:-label"
          aria-required="true"
          class="focus-visible:ring-ring focus-visible:ring-offset-background peer inline-flex h-4 w-8 shrink-0 cursor-pointer items-center rounded-full border-2 border-transparent transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 data-[state=checked]:bg-[#343839] data-[state=unchecked]:bg-[#D7D8D9] dark:data-[state=checked]:bg-[#3BBE5F] dark:data-[state=unchecked]:bg-[#404446]"
          data-state="checked"
          id=":r1:"
          role="switch"
          type="button"
          value="on"
        >
          <span
            class="peer pointer-events-none block size-3.5 rounded-full bg-[#FCFCFC] shadow-lg ring-0 transition-transform data-[state=checked]:translate-x-3.5 data-[state=unchecked]:translate-x-0"
            data-state="checked"
          />
        </button>
        <input
          aria-hidden="true"
          checked=""
          name="test"
          required=""
          style="transform: translateX(-100%); position: absolute; pointer-events: none; opacity: 0; margin: 0px; width: 0px; height: 0px;"
          tabindex="-1"
          type="checkbox"
          value="on"
        />
        <label
          class="inline-flex items-center text-sm font-medium leading-[22px] text-gray-900 dark:text-gray-400 group-hover:text-gray-900 dark:group-hover:text-slate-50 ml-2"
          for=":r1:"
          id=":r1:-label"
        >
          Test Field
          <span
            class="ml-1 text-gray-800 group-hover:text-gray-900 dark:text-gray-400 dark:group-hover:text-slate-50"
          >
            *
          </span>
          <button
            data-state="closed"
            type="button"
          >
            <div
              data-testid="icon-fallback"
              style="width: 16px; height: 16px;"
            />
          </button>
        </label>
      </div>
      <p
        class="mb-0 inline-flex items-center text-xs font-medium text-yellow-900 dark:text-yellow-900"
        data-type="warning"
      >
        Test warning
      </p>
    </div>
  </form>
</div>
`;
